src/herder/test/HerderTests.cpp

```diff
@@ -5803,3 +5803,117 @@ TEST_CASE("exclude transactions by operation type", "[herder]")
+    simulation->startAllNodes();
+
+    // Crank A and B until they're on ledger 2
+    simulation->crankUntil(
+        [&]() {
+            return A->getLedgerManager().getLastClosedLedgerNum() == 2 &&
+                   B->getLedgerManager().getLastClosedLedgerNum() == 2;
+        },
+        2 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+    // Return the number of entires in a node's scphistory table for the given
+    // ledger number
+    auto getNumSCPHistoryEntries = [&](Application::pointer node,
+                                       uint32_t ledgerNum) {
+        auto& db = node->getDatabase();
+        auto prep = db.getPreparedStatement(
+            "SELECT COUNT(*) FROM scphistory WHERE ledgerseq = :l");
+        auto& st = prep.statement();
+        st.exchange(soci::use(ledgerNum));
+        uint32_t count;
+        st.exchange(soci::into(count));
+        st.define_and_bind();
+        st.execute(true);
+        return count;
+    };
+
+    // A and B should have 2 entries in their scphistory table for ledger 2. C
+    // should have 0.
+    REQUIRE(getNumSCPHistoryEntries(A, 2) == 2);
+    REQUIRE(getNumSCPHistoryEntries(B, 2) == 2);
+    REQUIRE(getNumSCPHistoryEntries(C, 2) == 0);
+
+    // Get messages from A and B
+    HerderImpl* herderA = dynamic_cast<HerderImpl*>(&A->getHerder());
+    HerderImpl* herderB = dynamic_cast<HerderImpl*>(&B->getHerder());
+    REQUIRE(herderA);
+    REQUIRE(herderB);
+    std::vector<SCPEnvelope> AEnvs = herderA->getSCP().getLatestMessagesSend(2);
+    std::vector<SCPEnvelope> BEnvs = herderB->getSCP().getLatestMessagesSend(2);
+
+    // Pass A and B's messages to C
+    for (auto const& env : AEnvs)
+    {
+        C->getHerder().recvSCPEnvelope(env);
+    }
+    for (auto const& env : BEnvs)
+    {
+        C->getHerder().recvSCPEnvelope(env);
+    }
+
+    // Crank C until it is on ledger 2
+    simulation->crankUntil(
+        [&]() { return C->getLedgerManager().getLastClosedLedgerNum() == 2; },
+        2 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+    // Get messages from C
+    HerderImpl* herderC = dynamic_cast<HerderImpl*>(&C->getHerder());
+    REQUIRE(herderC);
+    std::vector<SCPEnvelope> CEnvs = herderC->getSCP().getLatestMessagesSend(2);
+
+    // Pass C's messages to A and B
+    for (auto const& env : CEnvs)
+    {
+        A->getHerder().recvSCPEnvelope(env);
+        B->getHerder().recvSCPEnvelope(env);
+    }
+
+    // Crank A and B until they're on ledger 3
+    simulation->crankUntil(
+        [&]() {
+            return A->getLedgerManager().getLastClosedLedgerNum() == 3 &&
+                   B->getLedgerManager().getLastClosedLedgerNum() == 3;
+        },
+        2 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+    // A and B should now have 3 entries in their scphistory table for ledger 2.
+    REQUIRE(getNumSCPHistoryEntries(A, 2) == 3);
+    REQUIRE(getNumSCPHistoryEntries(B, 2) == 3);
```

Comment:
Should we check that C does not have extra copies of messages? I might have missed something but I think this may be possible.

Reply:
Good idea to explicitly check this case. I added a test for this in d39400b30d1c195b1968a090fe6fb033a8752b08.